hbase-server/src/test/java/org/apache/hadoop/hbase/master/normalizer/TestSimpleRegionNormalizer.java

```diff
@@ -69,517 +78,347 @@
+    final TableName tableName = TableName.valueOf(name.getMethodName());
+    final List<RegionInfo> regionInfos = createRegionInfos(tableName, 4);
+    final Map<byte[], Integer> regionSizes =
+      createRegionSizesMap(regionInfos, 1, 1, 10, 10);
+    setupMocksForNormalizer(regionSizes, regionInfos);
+    assertEquals(Period.ofDays(7), normalizer.getMergeMinRegionAge());
+    assertThat(
+      normalizer.computePlansForTable(tableName),
+      everyItem(not(instanceOf(MergeNormalizationPlan.class))));
+
+    // have to call setupMocks again because we don't have dynamic config update on normalizer.
+    conf.unset(MERGE_MIN_REGION_AGE_DAYS_KEY);
+    setupMocksForNormalizer(regionSizes, regionInfos);
+    assertEquals(
+      Period.ofDays(DEFAULT_MERGE_MIN_REGION_AGE_DAYS), normalizer.getMergeMinRegionAge());
+    final List<NormalizationPlan> plans = normalizer.computePlansForTable(tableName);
+    assertThat(plans, not(empty()));
+    assertThat(plans, everyItem(instanceOf(MergeNormalizationPlan.class)));
+  }
+
+  @Test
+  public void testHonorsMergeMinRegionSize() {
+    conf.setBoolean(SPLIT_ENABLED_KEY, false);
+    final TableName tableName = TableName.valueOf(name.getMethodName());
+    final List<RegionInfo> regionInfos = createRegionInfos(tableName, 5);
+    final Map<byte[], Integer> regionSizes = createRegionSizesMap(regionInfos, 1, 2, 0, 10, 10);
+    setupMocksForNormalizer(regionSizes, regionInfos);
+
+    assertFalse(normalizer.isSplitEnabled());
+    assertEquals(1, normalizer.getMergeMinRegionSizeMb());
+    final List<NormalizationPlan> plans = normalizer.computePlansForTable(tableName);
+    assertThat(plans, everyItem(instanceOf(MergeNormalizationPlan.class)));
+    assertThat(plans, iterableWithSize(1));
+    final MergeNormalizationPlan plan = (MergeNormalizationPlan) plans.get(0);
+    assertEquals(regionInfos.get(0), plan.getFirstRegion());
+    assertEquals(regionInfos.get(1), plan.getSecondRegion());
+
+    conf.setInt(MERGE_MIN_REGION_SIZE_MB_KEY, 3);
+    setupMocksForNormalizer(regionSizes, regionInfos);
+    assertEquals(3, normalizer.getMergeMinRegionSizeMb());
+    assertThat(normalizer.computePlansForTable(tableName), empty());
+  }
 
-    Configuration configuration = HBaseConfiguration.create();
-    configuration.setInt(AbstractRegionNormalizer.HBASE_REGION_NORMALIZER_MIN_REGION_COUNT_KEY, 4);
-    when(masterServices.getConfiguration()).thenReturn(configuration);
+  // This test is to make sure that normalizer is only going to merge adjacent regions.
+  @Test
+  public void testNormalizerCannotMergeNonAdjacentRegions() {
+    final TableName tableName = TableName.valueOf(name.getMethodName());
+    // create 5 regions with sizes to trigger merge of small regions. region ranges are:
+    // [, "aa"), ["aa", "aa1"), ["aa1", "aa1!"), ["aa1!", "aa2"), ["aa2", )
+    // Region ["aa", "aa1") and ["aa1!", "aa2") are not adjacent, they are not supposed to
+    // merged.
+    final byte[][] keys = {
```

Comment:
That is a nice move, thank you! I tested without the patch(sort based regioninfo in getMergePlan()), it also passed for me. Debugged a bit, regionInfos() here is already sorted based on STARTKEY. We need to simulate what is the order in master, so changing the keys to 
`    final byte[][] keys = {
      null,
      Bytes.toBytes("aa"),
      Bytes.toBytes("aa1!"),
      Bytes.toBytes("aa1"),
      Bytes.toBytes("aa2"),
      null,
    };`
will fix it. The regionInfo is already key based, the order does not matter.

Reply:
Oh, got it. Makes sense, thanks.

Reply:
Or, sorry, there is another change needed.
    final Map<byte[], Integer> regionSizes = createRegionSizesMap(regionInfos, 3, 1, 1, 3, 5);
Since the order of regions are changed, the regionSizes order also needs to change.

Reply:
`diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/master/normalizer/TestSimpleRegionNormalizer.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/master/normalizer/TestSimpleRegionNormalizer.java
index 6b75ed7881..eeb9bef0c8 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/master/normalizer/TestSimpleRegionNormalizer.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/master/normalizer/TestSimpleRegionNormalizer.java
@@ -396,13 +396,13 @@ public class TestSimpleRegionNormalizer {
     final byte[][] keys = {
       null,
       Bytes.toBytes("aa"),
-      Bytes.toBytes("aa1"),
       Bytes.toBytes("aa1!"),
+      Bytes.toBytes("aa1"),
       Bytes.toBytes("aa2"),
       null,
     };
     final List<RegionInfo> regionInfos = createRegionInfos(tableName, keys);
-    final Map<byte[], Integer> regionSizes = createRegionSizesMap(regionInfos, 3, 1, 3, 1, 5);
+    final Map<byte[], Integer> regionSizes = createRegionSizesMap(regionInfos, 3, 1, 1, 3, 5);
     setupMocksForNormalizer(regionSizes, regionInfos);
 
     // Compute the plan, no merge plan returned as they are not adjacent.`